src/nvim/ex_docmd.c

```diff
@@ -2423,9 +2423,14 @@ static char_u *find_command(exarg_T *eap, int *full)
       while (ASCII_ISALNUM(*p))
         ++p;
 
-    /* check for non-alpha command */
-    if (p == eap->cmd && vim_strchr((char_u *)"@!=><&~#", *p) != NULL)
-      ++p;
+    // check for non-alpha command
+    if (p == eap->cmd) {
+      if (p[0] == '&' && p[1] == ':') {
+        p = p + 2;
+      } else if (vim_strchr((char_u *)"@!=><&~#", *p) != NULL) {
+        p++;
+      }
+    }
```

Comment:
Other way to handle ambiguity:

- add a new function like `ex_sub_or_async` which decides whether to call `ex_substitute` or `ex_async`. 
  -  use `skipwhite()`  to check if the next char is `:`. 
- in `ex_cmds.lua`, 
  - set `func='ex_sub_or_async'` for the `&` command.
  - add `NOTRLCOM`


Not sure if that's better, can just leave it for now.

Reply:
Hmmm...I didn't want to change the `&` command mainly because semantically our new async command is not a special case of substitution.

Concerning `&:vimgrep` handler, I need to modify it so that it can handle paths more properly just like `vimgrep`, currently it just search `glob(path, v:true, v:true)` of whatever path given, so paths like `**/*.h **/*.c` will not work unlike `vimgrep`, not a big deal to parse I guess.

Not sure about the next step, maybe I'll add similar async equivalents to `vimgrepadd`, `lvimgrep`, and `lvimgrepadd` for now.

I'm open to any kind of feedback, once we settle we can start working on the more interesting stuff like context handling and async calls of user-defined functions (which could be just making the called function part of the sent context and then call it as usual).

Reply:
> Not sure about the next step, maybe I'll add similar async equivalents to `vimgrepadd`, `lvimgrep`, and `lvimgrepadd` for now.

Yes, implementing a few more cases will help to start seeing a pattern (and see what's painful/boilerplate/brittle). Then we can start to form a best-practice for adding this functionality.